prototypes/emdpg_waveguide/EM_waveguide.cc

```diff
@@ -0,0 +1,2451 @@
+
+    add_parameter("n repetitions",
+                  n_repetitions,
+                  "Number of repetitions for the subdivided hyper rectangle");
+
+    add_parameter("adaptive refinement",
+                  adaptive_refinement,
+                  "If true, the code will use adaptive mesh refinement, "
+                  "otherwise it will use uniform mesh refinement");
+
+    add_parameter("max refinement level",
+                  max_refinement_level,
+                  "Maximum number of refinements from the coarse mesh");
+
+    // Default finite element parameters
+    degree = 0;
+    add_parameter("degree", degree, "order of the DGQ finite element space");
+
+    delta_degree = 1;
+    add_parameter("delta degree",
+                  delta_degree,
+                  "order difference between the trial and test space");
+  }
+
+  /* Create analytical solution class for the electric field (E).
+   * The corresponding solution for the electric field follows the physicist's
+   * time convention with the imaginary unit in the exponent being negative
+   * (i.e. exp(-i wt) ). The function manage both the real and complex
+   * components of the solution. Since the frequency, the material properties,
+   * and the geometry are arguments for the parameter file, they are used in the
+   * constructor to initialize the different wavenumber. Note that everything is
+   * addimensionalize using the following : \tilde{\mathbf{E}} =
+   * \frac{\mathbf{E}}{E_0}, \tilde{\mathbf{H}} = \frac{Z_0}{E_0} \mathbf{H},
+   *       \varepsilon_r = \frac{\varepsilon}{\varepsilon_0},
+   *       \mu_r = \frac{\mu}{\mu_0},
+   *       Z_r = \frac{Z}{Z_0},
+   *       \sigma_r = \frac{\sigma}{\omega \varepsilon_0},
+   *       \tilde{\omega} =  \frac{\omega}{k_0 c_0},
+   *       \tilde{\mathbf{J}} = \frac{1}{\omega \varepsilon E_0 } \mathbf{J}
+   *       \tilde{k} = L \mathbf{k},
+   *       \tilde{\nabla} = L \nabla.
+   */
+
+  template <int dim>
+  class AnalyticalSolution_E : public Function<dim>
+  {
+  public:
+    // Constructor
+    AnalyticalSolution_E(const Parameters<dim> &parameters)
+      : Function<dim>(6)
+      , parameters(parameters)
+      , k(2. * M_PI * parameters.frequency / parameters.speed_of_light)
+      , k_x(parameters.mode_x * M_PI / parameters.waveguide_a)
+      , k_y(parameters.mode_y * M_PI / parameters.waveguide_b)
+      , k_z(std::sqrt(k * k - k_x * k_x - k_y * k_y))
+      , k_c(std::sqrt(k_x * k_x + k_y * k_y))
+      , Z_r(std::sqrt(parameters.mu_r / parameters.epsilon_r))
+    {}
+
+    virtual double
+    value(const Point<dim> &p, const unsigned int component) const override;
+
+  private:
+    const Parameters<dim>     &parameters;
+    const double               k;
+    const double               k_x;
+    const double               k_y;
+    const double               k_z;
+    const double               k_c;
+    const std::complex<double> Z_r;
+  };
+
+  template <int dim>
```

Comment:
I would add a comment here to the effect that the solution for E is
Real(E_x), Real(E_y), Real(E_z), Im(E_x), Im(E_y), Im(E_z).

This makes sense, but it is worth to repeat it.